lib/SPIRV/SPIRVReader.cpp

```diff
@@ -382,11 +382,18 @@ Type *SPIRVToLLVM::transType(SPIRVType *T, bool IsClassMember) {
   case OpTypeArray:
     return mapType(T, ArrayType::get(transType(T->getArrayElementType()),
                                      T->getArrayLength()));
-  case OpTypePointer:
+  case OpTypePointer: {
+    SPIRAddressSpace AS =
+        SPIRSPIRVAddrSpaceMap::rmap(T->getPointerStorageClass());
+    // If generation of SYCL USM address spaces isn't allowed in the reversed
+    // translation - transform them into global address space
+    if (!BM->isUsmAddrspacesEnabled() &&
```

Comment:
What is the reason to add this option? I mean, new storage classes enable `USMStorageClassesINTEL` capability in the module and it is responsibility of the particular consumer to either reject SPIR-V with such capability as unsupported or accept new storage classes, which can be done by lowering them to whichever address spaces or to anything else what is desired by the consumer.

So, the generator is responsible to either generate more portable SPIR-V, which doesn't use the extension, or generate less portable SPIR-V, which uses the extension, but cannot be supplied to any consumer (because of lack of the extension support).

If consumer is aware of that option, it means that it is aware of that extension and therefore, it could perform desired lowering by itself and code in the translator would be simpler.

I do understand that some projects are using the translator as SPIR-V consumer as well and this option allows to simplify their code, but it would be good if such customizations are not unique to a single consumer/backend/vendor and we don't end up with bunch of tweaks in the translator for each particular backend.

P.S. This is just my vision and it might be wrong. Probably it is such a problem to have some consumer part customizations not in the actual SPIR-V consumers, but here, in the translator. It would be good to hear feedback from someone else

Reply:
In general I don't see such customization as a big problem, we have here in the translator different flows for OpenCL 2.0 and OpenCL 1.2, right? Yet I see your point and understand it. Would also love to here other opinions.

(The story below isn't valuable for SPIR-V to LLVM IR translator project at all, but I want you to see the logic, why this option was added).
This option came from (surprise-surprise) some implementation details of SYCL compiler's driver and desire to make the feature adoptable not only for FPGA, but for other targets in JIT compilation (aka with triple `spirv-unknown-unknown` for which we add `spirv-ext=+all` option.

Reply:
> we have here in the translator different flows for OpenCL 2.0 and OpenCL 1.2, right?

This is for sure shared between several vendors

> This option came from (surprise-surprise) some implementation details of SYCL compiler's driver and desire to make the feature adoptable not only for FPGA, but for other targets in JIT compilation (aka with triple spirv-unknown-unknown for which we add spirv-ext=+all option.

This is exactly what I'm talking about: device-/vendor-specific consumption rules are being moved into the translator. I think it is okay to outline common parts into the translator, but at the same time I wouldn't like to see bunch of vendor-specific options in this repo.

I don't have strong objections at this point: probably we can add this option as a first example and when we have the second one we can see how this can be generalized, where it goes to and how does it look like

Reply:
A SPIR-V module declares capabilities, i.e. features which are used in the module. A client API environment (aka target environment) should check capabilities declared in the module against capabilities it supports. If the client API doesn't support a capability declared in the module it's "allowed to reject the module."
The `EnableUsmAddrspaces` option looks like we are trying adapt a SPIR-V module with `USMStorageClassesINTEL` capability for client API(device) which doesn't support that capability. IMO a client API should **either support the capability or reject the module**.
I agree with @AlexeySachkov, we should not move device specific logic into the translator.

Reply:
Ok. Do you have any comments other than this one?